src/dag.rs

```diff
@@ -0,0 +1,483 @@
+        assert!(already_inserted.is_none());
+    }
+
+    pub fn get(
+        &self,
+        name: &ActionName,
+    ) -> Result<Arc<dyn Action<UserType>>, ActionRegistryError> {
+        self.actions.get(name).cloned().ok_or(ActionRegistryError::NotFound)
+    }
+}
+
+/// A Node in the graph
+///
+/// Since sagas are constructed dynamically at runtime, we don't know the
+/// shape of the graph ahead of time. We need to maintain enough information
+/// to reconstruct the saga when loaded from persistent storage. The easiest
+/// way to do that is to store the graph itself with enough information in
+/// each node that allows us to recreate the Saga. Note that we don't store
+/// the execution state of the saga here. That continues to reside in saga log
+/// consisting of `SagaNodeEvent`s.
+///
+/// There can be multiple subsagas with nodes that run the same actions. In order
+/// to distinguish the action outputs we tag each one with an `instance_id`.
+#[derive(Debug, Clone, Serialize, Deserialize, JsonSchema)]
+pub struct Node {
+    pub name: String,
+    pub instance_id: u16,
+    pub label: String,
+    pub action: ActionName,
+
+    /// Each node may be the start of the outer saga, or a subsaga. We store
+    /// the creation parameters for the saga or subsaga as output of the first
+    /// node so that it can be retrieved by child nodes. In order to feed the
+    /// output of the node execution, the params themselves must be stored
+    /// as input to the DAG, and so we store them here. We only store these
+    /// parameters for the first node of a saga or child saga.
+    pub create_params: Option<serde_json::Value>,
+    // TODO:
+    // What if we hang a set of KV pairs or other structured data off each node
+    // which instructs that action what lookup keys to use? This could
+    // be instead of `instance_id`. The benefits are more information
+    // could be dynamically added to nodes in the DagBuilder. In essence, each node
+    // could get its own set of "create parameters".
+    // The downside is more work for creators of dags and users of actions. This
+    // would also likely be more error prone in that now each user of a dag
+    // has to add the right data structure to each node which runs an action or
+    // the action won't do the right thing. So instead of depending just on the output
+    // of node actions, the actions also depend on input of their current node.
+    // If we did this, could we typecheck these inputs with the actions? Maybe via
+    // some macro that generates a checker function.
+}
+
+/// A specification for creating a node. This allows dynamic creation of nodes
+/// for subsagas with different instance_ids and creation parameters.
+pub struct NodeSpec<'a> {
+    pub name: &'a str,
+    pub label: &'a str,
+    pub action: &'a str,
+}
+
+/// An abstract description of the shape of node execution
+///
+/// TODO: Do we want to allow arbitrary recursion here?
+pub enum NodeConcurrency<'a> {
```

Comment:
Do we want to allow arbitrary shapes/recursion here? This currently matches the `append`/`append_paralle`l API in the DagBuilder (and previously `SagaTemplateBuilder`)